packages/report-server/src/reportBuilder/transform/functions/insertRows.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup } from '@tupaia/utils';
+
+import { TransformParser } from '../parser';
+import { functions } from '../../functions';
+import { buildWhere } from './where';
+import { Row } from '../../types';
+import { getParsedColumnKeyAndValue } from './helpers';
+import { mapStringToStringValidator } from './transformValidators';
+
+type InsertParams = {
+  columns: { [key: string]: string };
+  where: (parser: TransformParser) => boolean;
+  positioner: (index: number, insertCount: number) => number;
+};
+
+const positioners = {
+  before: (index: number, insertCount: number) => index + insertCount,
+  after: (index: number, insertCount: number) => index + insertCount + 1,
+  start: (index: number, insertCount: number) => insertCount,
+};
+
+const paramsValidator = yup.object().shape({
+  columns: mapStringToStringValidator,
+  where: yup.string(),
+  position: yup
+    .mixed<'before' | 'after' | 'start'>()
+    .oneOf(['before', 'after', 'start'])
+    .default('after'),
+});
```

Comment:
Made the switch to `yup` for all input validation. Makes code a LOT more readable, and typescript works really nice with it. However, have a slight concern about the performance implications...